File: examples/ppt-replace.ts
/**
 * 例子：打开一份含 `{{date}}` 占位的 pptx，把全部占位替换为指定字符串，写出。
 *
 * 跑法：
 *   bun run examples/ppt-replace.ts <input.pptx> <output.pptx> <replacement>
 */

import { Text } from "../src/drawing/index.js";
import { PresentationDocument } from "../src/ppt/index.js";

async function main(): Promise<void> {
  const [inputPath, outputPath, replacement] = process.argv.slice(2);
  if (inputPath === undefined || outputPath === undefined || replacement === undefined) {
    process.stderr.write("usage: ppt-replace <input.pptx> <output.pptx> <replacement>\n");
    process.exit(2);
  }

  const doc = await PresentationDocument.openAsync(inputPath);
  const pp = doc.presentationPart;
  if (pp === undefined) {
    process.stderr.write("error: input has no presentation part\n");
    process.exit(1);
  }

  let replacedCount = 0;
  for (const sp of pp.slideParts) {
    for (const t of sp.slide.descendants(Text)) {
      if (t.text?.includes("{{date}}")) {
        t.text = t.text.replaceAll("{{date}}", replacement);
        replacedCount += 1;
      }
    }
  }

  await doc.saveAsAsync(outputPath);
  process.stdout.write(`Replaced ${replacedCount} occurrence(s); wrote ${outputPath}\n`);
}

main().catch((err) => {
  process.stderr.write(`failed: ${(err as Error).stack ?? err}\n`);
  process.exit(1);
});

File: tools/cross-sdk-verify/replicas/PptReplace.cs
// Replica of examples/ppt-replace.ts
// The TS runner generates a ppt-create output (3 slides, slide 2 has "Generated on {{date}}")
// then calls ppt-replace.ts with replacement "2024-01-01".
// This replica replicates that pipeline: build the 3-slide ppt-create template, then
// replace every "{{date}}" occurrence with "2024-01-01".

using DocumentFormat.OpenXml.Packaging;
using A = DocumentFormat.OpenXml.Drawing;

namespace CrossSdkVerify.Replicas;

public static class PptReplace
{
    private const string Replacement = "2024-01-01";

    public static void Run(string outputPath)
    {
        // Step 1: build the ppt-create 3-slide template into a temp file
        var tempPath = Path.GetTempFileName() + ".pptx";
        try
        {
            PptCreate.Run(tempPath);
            ApplyReplace(tempPath, outputPath);
        }
        finally
        {
            if (File.Exists(tempPath)) File.Delete(tempPath);
        }
    }

    private static void ApplyReplace(string inputPath, string outputPath)
    {
        File.Copy(inputPath, outputPath, overwrite: true);

        using var doc = PresentationDocument.Open(outputPath, true);
        var presPart = doc.PresentationPart;
        if (presPart == null) return;

        foreach (var slidePart in presPart.SlideParts)
        {
            foreach (var t in slidePart.Slide.Descendants<A.Text>())
            {
                if (t.Text?.Contains("{{date}}") == true)
                    t.Text = t.Text.Replace("{{date}}", Replacement);
            }
        }

        presPart.Presentation.Save();
    }
}
